test/test_torch.py

```diff
@@ -6293,6 +6293,75 @@ def test_cat_different_devices(self, devices):
                                     "input tensors must be on the same device"):
             torch.cat((cpu, cuda0))
 
+    def test_block_diag(self, device):
+        def block_diag_workaround(*arrs):
+            arrs_expanded = []
+            for a in arrs:
+                if a.dim() == 2:
+                    arrs_expanded.append(a)
+                elif a.dim() == 1:
+                    arrs_expanded.append(a.expand(1, a.size(0)))
+                elif a.dim() == 0:
+                    arrs_expanded.append(a.expand(1, 1))
+            shapes = torch.tensor([a.shape for a in arrs_expanded], device=device)
+            out = torch.zeros(
+                torch.sum(shapes, dim=0).tolist(),
+                dtype=arrs_expanded[0].dtype,
+                device=device
+            )
+            r, c = 0, 0
+            for i, (rr, cc) in enumerate(shapes):
+                out[r:r + rr, c:c + cc] = arrs_expanded[i]
+                r += rr
+                c += cc
+            return out
+
+        tensors = [
+            torch.rand((2, 2), device=device),
+            torch.rand((2, 3), device=device),
+            torch.rand(10, device=device),
+            torch.rand((8, 1), device=device),
+            torch.rand(1, device=device)[0]
+        ]
+        result = torch.block_diag(*tensors)
+        result_check = block_diag_workaround(*tensors)
+        self.assertEqual(result, result_check)
+
+        tensor = torch.rand(1, device=device)[0]
+        result = torch.block_diag(tensor)
+        result_check = tensor.expand(1, 1)
+        self.assertEqual(result, result_check)
+
+        tensor = torch.rand(10, device=device)
+        result = torch.block_diag(tensor)
+        result_check = tensor.expand(1, tensor.size(0))
+        self.assertEqual(result, result_check)
+
+    @unittest.skipIf(not TEST_SCIPY, "Scipy not found")
+    def test_block_diag_scipy(self, device):
+        import scipy.linalg
+        scipy_tensors = [
+            1,
+            [2],
+            [],
+            [3, 4, 5],
+            [[], []],
+            [[6], [7.3]]
+        ]
+        torch_tensors = [torch.tensor(t, device=device) for t in scipy_tensors]
+        torch_result = torch.block_diag(*torch_tensors)
+        self.assertEqual(torch_result.dtype, torch.float32)
+
+        scipy_result = torch.tensor(
+            scipy.linalg.block_diag(*scipy_tensors),
+            device=device
+        )
+
+        if scipy_result.dtype == torch.float64:
```

Comment:
If I remember correctly, no. One of the CI jobs was getting a float32 here.

Reply:
Oh actually I'm wrong. The problem was that the scipy result is float64, but the torch result is float32. Pytorch seems to have a policy of preferring 32-bit to 64-bit. I added this conditional so that if someone adds more of these scipy test cases (like I am about to do), then we can add more conditionals for different types, like complex128 -> complex64. Maybe not the best way to do it though, what do you think?

Reply:
Maybe I should demote the scipy result's datatype rather than promote the torch result type.

Reply:
I just added a few more cases to this test, changing how I was dealing with these precision differences. Let me know what you think.

Reply:
Those look great to me